grammar.js

```diff
@@ -498,7 +496,16 @@ module.exports = grammar({
     ),
     _version_vector: $ => /\$[a-zA-Z_][a-zA-Z0-9-_]*\$/,
 
-    uri: $ => /\w+:(\/?\/?)[^\s}]+/,
+    uri: $ => /https?:\/\/[^\s}]+/,
+
+    fqsen: $ => choice(
+      $.named_type,
+      $.variable_name,
+      seq($.name, '()'),
+      seq($.named_type, '::', $.name),
+      seq($.named_type, '::', $.name, '()'),
+      seq($.named_type, '::', $.variable_name)
```

Comment:
Here I wondered if we should reuse things like [scoped_property_access_expression](https://github.com/tree-sitter/tree-sitter-php/blob/0ce134234214427b6aeb2735e93a307881c6cd6f/grammar.js#L1027-L1031) or [class_constant_access_expression](https://github.com/tree-sitter/tree-sitter-php/blob/0ce134234214427b6aeb2735e93a307881c6cd6f/grammar.js#L882-L886). But this would result in a quite different node structure. And in our case it's not really an access either. So I went with a definition similar to types.

Reply:
So, I *think* that I would prefer to explicitly share/reuse the PHP parser nodes in this case. The upside of doing so, in my limited experience, is that it means that we can have grammar queries that match the same syntax "things" whether they appear in code or in docblocks, which has been nice. eg by reusing the type definitions where we parse types for return/param/etc has meant that my editor has started making those clickable w/o any add'l config or effort, ie just by happy accident. Reusing the php parser nodes here might also make the same work for `@see`. 

Unless I'm mistaken, the 3 class access nodes we're looking for are `class_constant_access_expression`, `scoped_property_access_expression` and `scoped_call_expression`. I wonder if `function_call_expression` would cover the bases for L504, or if something else might be needed there.

Please advise if you think going this route would be unwise, though. Perhaps you can elaborate more on why you went this route vs reusing PHP nodes?

Reply:
The problem I see here is that the PHP rules match more than what we need here and thus are not really apropriate. For example `scoped_property_access_expression` uses `_scope_resolution_qualifier` which uses  [_dereferencable_expression](https://github.com/tree-sitter/tree-sitter-php/blob/0ce134234214427b6aeb2735e93a307881c6cd6f/grammar.js#L1150-L1159). I think it covers expressions like `$className::someMethod()` which can not appear in a docblock. Like I said: Syntactically we don't deal with accessing a prop/constant here which has way more options than what we need here.

See my other answer below for my thoughts on PHP rule reuse.


> we can have grammar queries that match the same syntax "things" whether they appear in code or in docblocks
> [...]  my editor has started making those clickable w/o any add'l config or effort, ie just by happy accident

But isn't this up to us, which queries we use for phpdoc? We don't have to reuse the PHP queries. Thinking more about it, are you sure that this happy accident is really due to treesitter? At least in nvim this is covered by LSP and treesitter is only used for syntax highlighting.